### ISSUE TITLE
FreeSWITCH sofia profile reverse proxy IPv4/IPv6 mismatch

### ISSUE THREAD

**Issue by TheJJ** [bigbluebutton/bigbluebutton#9295]
The symptom is that joined meetings randomly fail with error `1004 INCOMPATIBLE SDP`.
The failure occurs for all browsers, and when meetings are refreshed, it often works.
If you refresh often enough, it will work for sure eventually.

This happens as freeswitch refuses the offered IPs in ICE with `no network path`.
Then `no suitable candidates found`, and hangs up with `INCOMPATIBLE_DESTINATION`.

The underlying cause is this:
The SIP-websocket is reverse-proxied with nginx to freeswitch with `/etc/bigbluebutton/nginx/sip.nginx`. There, the freeswitch IP or hostname is entered.
The proxied connection decides which sofia profile freeswitch will select, since the freeswitch has two sofia profiles (external and external-ipv6). And this profile determines, how and if a connection to the client (the bbb web client) can happen.

* If one does `proxy_pass https://v4address:7443;`, freeswitch will always select the ipv4 sofia profile, and thus only talk to clients with v4, even if they offer v6 addresses through ICE.
* If one does `proxy_pass https://v6address:7443;`, freeswitch will select the `v6` sofia profile, and refuse to talk with `v4` only clients! This is because the `rtpip` in the profile is a `v6` only, and during ICE freeswitch can't match the offered `v4` addresses with its `v6`-only sip profile address.
* If one enters a hostname (`proxy_pass https://node9001.bbb.myuni.edu`) that has both `v4` and `v6`, the selected sofia profile will be quite random, and client connections will fail if the `v6`-profile is selected but the client doesn't have a v6 address offering in ICE.

Solution:
Keep the `v4`/`v6` path in tact, through the reverse proxy.

Define this at nginx config top level (above the `server` block):
``` nginx
geo $client_is_v6 {
    ::/0           true;
}

# here, the bbb nginx server is defined:
server {
    listen 443 ssl http2;
    listen [::]:443 ssl http2;
    ...
```

In `sip.nginx`, the `client_is_v6` selects the v6 or v4 address of freeswitch:
``` nginx
# redirect to freeswitch wss listener
location /ws {
        proxy_pass https://123.456.789.012:7443;

        if ($client_is_v6) {
            proxy_pass https://[dead:beef::1337]:7443;
        }
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        proxy_read_timeout 6h;
        proxy_send_timeout 6h;
        client_body_timeout 6h;
        send_timeout 6h;
}
```

This fixes the correct sofia profile selection, depending on client connectivity.
The sporadic 1004 incompatible sdp errors are gone.

When you tell me where the `sip.nginx` file is (I can't find it..), I can create a pull to patch this issue. 

TUM [in](https://in.tum.de) and [ma](https://ma.tum.de) thank you for BBB :)

**Comment by kepstin** [comment]
I actually use a slightly different solution to this problem, which avoids the (sometimes problematic) use of "if" statements in the nginx config.

Create the file `/etc/nginx/conf.d/bigbluebutton_sip_addr_map.conf` with the contents:
```
map $remote_addr $freeswitch_addr {
	"~:"	[2001:db8::1];
	default	192.0.2.1;
}
```

And then modifying the `proxy_pass` line in `/etc/bigbluebutton/sip.nginx` to read
```
proxy_pass https://$freeswitch_addr:7443;
```

**Comment by synnack** [comment]
Just wanted to let you know I've hit this bug too, and the fix from @kepstin is pretty clean and it works great, thanks!

**Comment by znerol** [comment]
This is already documented in [the troubleshooting section](http://docs.bigbluebutton.org/2.2/troubleshooting.html), Scroll down to "Configure BigBluebutton/FreeSWITCH to support IPV6"

**Comment by TheJJ** [comment]
Great! I'd like to fix this properly in the default configuration, though.

**Comment by synnack** [comment]
Seems more useful to be fixed in the installer, not the documentation. 

There should be no scenarios where including this will fail, you won't come in on IPv4 when IPv4 is unsupported and won't come in on IPv6 when IPv6 is unsupported.

**Comment by steffen-moser** [comment]
I'd like to comment on this issue, as I am not sure if it is fixed fo sure. 

We do see some 1004 errors since we've activated IPv6 in our BBB infrastructure. We have not found out how they occur, but as far as we know yet, the following facts play a role in it: 

 * The user seems to have a control connection using TCP/IPv6, i.e. they load the BBB client and the control channel using IPv6.
 * When sending the client's SDP headers, it only contains IPv4 addresses as candidates. 
 * In this case, FreeSwitch seems to come to the conclusion that a connection is not possible as FreeSwitch has been put into the IPv6 world (due to the fact, that the control connection has been running on IPv6). You'll also find that FreeSwitch is going to offer an IPv6 address only.

We still do not know what triggers the client (we saw it on Firefox 72.0.1 and on Chrome, both running on OpenSUSE Linux 42.3, but we were not able to reproduce it in our test lab) to propose only IPv4 addresses as candidates in the SDP header. 

At the same time, we found that blocking all UDP/IPv6 ports on a user's router (by using NETFILTER/ip6tables rules) on purpose, we can provoke similar problems. In this case, the control connection is established via IPv6, but there won't be any UDP datagrams going through on IPv6. In this case, the client might also propose only IPv4 addresses, which can lead so similar problems. Whether a router or a router's firewall issues are the cause for the above-described problem, has not been completely worked out, yet. It seems to be one possible way to provoke a maybe similar problem.

My questions are: 
* Has anybody seen problems like this? 
* Is there any chance to have FreeSwitch to behave more like Kurento here? In my opinion, a behavior where addresses exchanged during the SDP procedure will be independent of a previous control channel's network stack, would lead to the least trouble.

**Comment by znerol** [comment]
> We still do not know what triggers the client (we saw it on Firefox 72.0.1 and on Chrome, both running on OpenSUSE Linux 42.3, but we were not able to reproduce it in our test lab) to propose only IPv4 addresses as candidates in the SDP header.

I've observed that some browsers+OS combinations will exclusively advertise candidates detected via STUN/TURN. I.e., they only advertise `srlfx` and `relay` but no single `host` [candidate](https://developer.mozilla.org/en-US/docs/Web/API/RTCIceCandidateType). It follows that your STUN/TURN server must be dual-stack, otherwise some clients will only advertise candidates for one IP protocol (IPv4).

**Comment by steffen-moser** [comment]
I totally agree with you regarding the browsers' strange behavior, we still do not know why this happens in all cases. Maybe some firewalls, maybe some security-enhancing or privacy-improving browser plugins/extensions also play some roles. Maybe these browser plugins/extensions are active system-wide and cannot be found in the user's browser profile, maybe some personal security suites (e.g. former commercial virus scanning software) are introducing this kind of trouble...

Your conclusion partly coincides with our observation. There might still be clients, which have a working TCP/IPv6, but a firewall blocking UDP/IPv6*. In this case, browsers might offer IPv4 addresses only, and FreeSwitch ends with 1004 as it has switched to an IPv6-only mode itself - despite being in a dual-stack setup. The problem is: FreeSwitch acts only dual-stack-like as long as no web-socket management connection has been made. As soon as this happened, the stack to use for WebRTC is nailed to IPv4 or IPv6, either. 

My colleague has just described the problem in a very detailed way here: https://groups.google.com/forum/#!topic/bigbluebutton-setup/OzjwjFe7Ksk

Kurento applications (listen to audio-only, video, screen sharing) don't seem to be this fussy. There might also be corner cases where firewall problems cause trouble, but we've seen situations where "audio-only" (handled by Kurento) worked fine, but bi-directional audio (handled by FreeSwitch) caused 1004 and 1007 errors.

My suggestions/questions are: Is there any chance to loosen FreeSwitch's behavior of strictly deciding the WebRTC stack profile based on the IP stack used in the previous web-socket connection? I could at least imagine that a less strict behavior might improve the user-experience - at least for some of the users - in a dual-stack setup. At the same time, I do not see any negative impact. 


*I suspect that strange firewall settings in small-office/home-office routers especially for IPv6 are not this rare. While in the IPv4 era it was common to have one single public IP address, SOHO routers were often configured under the (wrong) assumption that NAT is something like a firewall and the internal clients and their accidentally offered services are not reachable from the outside (STUN's abilities easily show that this is not true). In IPv6, where the clients' IP addresses are exposed publicly, router manufacturers start to deal with sophisticated firewall rules, some of them might be more and some might be less reasonable, but they all might affect WebRTC...

**Comment by tacerus** [comment]
I wonder, how do you configure FreeSWITCH to have the WebSockets listen on both, IPv4 and IPv6? For some reason, this configuration

```
    <param name="ws-binding" value=":5066"/>
    <param name="wss-binding" value=":7443"/>
```

binds the ports only to the IPv6 address of the main network interface.

**Comment by tacerus** [comment]
Figured it out - I have been under the false assumption that FreeSwitch supports multiple listeners per profile. After configuring the same bindings in `/etc/freeswitch/sip_profiles/external-ipv6.xml` and defining `local_ip_v6` in `vars.xml`

`<X-PRE-PROCESS cmd="set" data="local_ip_v6=::1"/>`

the service started to bind with dual stack.
